# Beginning Our Data-driven Journey in Maji Ndogo

## Introduction

In this first part of the integrated project, we dive into Maji ndogo's expansive dataset containing just over 60000 records spread across various tables. As we navigate this trove of data, we'll use basic queries to familiarise ourselves with the contents of each table in the database. We'll also use SQL **Data Manipulation Language (DML)** to refine some data points while we're at it.

## Notebook Setup

In [1]:
# Load the sql extension
%load_ext sql

In [2]:
# Create a connection to the mysql 'md_water_services' database
%sql mysql+pymysql://root:password@localhost:3306/md_water_services

Connecting to 'mysql+pymysql://root:***@localhost:3306/md_water_services'

## Familiarising Ourselves With the Data

Let's start by reviewing the first few records of each table to get a high level overview of what our data looks like. First things first, let's see the tables that are in Maji Ndogo's database.

In [3]:
%sql SHOW TABLES

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

8 rows affected.

+-----------------------------+
| Tables_in_md_water_services |
+-----------------------------+
|       data_dictionary       |
|           employee          |
|     global_water_access     |
|           location          |
|            visits           |
|        water_quality        |
|         water_source        |
|        well_pollution       |
+-----------------------------+

We can see that we have a total of **8** tables. Let's see what each of these tables contain starting with the `data_dictionary` table.

In [4]:
%sql SELECT * FROM data_dictionary;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

49 rows affected.

+------------+----------------------+-------------------------------------------------+--------------+-----------------------------+
| table_name |     column_name      |                   description                   |   datatype   |          related_to         |
+------------+----------------------+-------------------------------------------------+--------------+-----------------------------+
|  employee  | assigned_employee_id |       Unique ID assigned to each employee       |     INT      |            visits           |
|  employee  |    employee_name     |               Name of the employee              | VARCHAR(255) |                             |
|  employee  |     phone_number     |          Contact number of the employee         | VARCHAR(15)  |                             |
|  employee  |        email         |          Email address of the employee          | VARCHAR(255) |                             |
|  employee  |       address        |       Residential address of the employee       | VARCHAR(255) |                             |
|  employee  |      town_name       |   Name of the town where the employee resides   | VARCHAR(255) |                             |
|  employee  |    province_name     | Name of the province where the employee resides | VARCHAR(255) |                             |
|  employee  |       position       |      Position or job title of the employee      | VARCHAR(255) |                             |
|   visits   |      record_id       |         Unique ID assigned to each visit        |     int      | water_quality, water_source |
|   visits   |     location_id      |            ID of the location visited           | varchar(255) |           location          |
+------------+----------------------+-------------------------------------------------+--------------+-----------------------------+
Truncated to displaylimit of 10.

We notice that the data dictionary has description of column names per table in the database. So to get any information a specific table and their column names along with description of each column we can just run a query like below.

In [5]:
%sql SELECT column_name, description, datatype, related_to FROM data_dictionary WHERE table_name = "employee";

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

8 rows affected.

+----------------------+-------------------------------------------------+--------------+------------+
|     column_name      |                   description                   |   datatype   | related_to |
+----------------------+-------------------------------------------------+--------------+------------+
| assigned_employee_id |       Unique ID assigned to each employee       |     INT      |   visits   |
|    employee_name     |               Name of the employee              | VARCHAR(255) |            |
|     phone_number     |          Contact number of the employee         | VARCHAR(15)  |            |
|        email         |          Email address of the employee          | VARCHAR(255) |            |
|       address        |       Residential address of the employee       | VARCHAR(255) |            |
|      town_name       |   Name of the town where the employee resides   | VARCHAR(255) |            |
|    province_name     | Name of the province where the employee resides | VARCHAR(255) |            |
|       position       |      Position or job title of the employee      | VARCHAR(255) |            |
+----------------------+-------------------------------------------------+--------------+------------+

The information above tells us that the `employee` table has **8** columns on of which seems to be a primary key related to another table i.e. `assigned_employee_id` is used to reference some information in the `visits` table. We can even retrieve table names that are related to each other by running a query like so. 

In [6]:
%%sql
# Retrieve related tables
SELECT DISTINCT table_name
FROM data_dictionary
WHERE related_to != "";

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

6 rows affected.

+----------------+
|   table_name   |
+----------------+
|    employee    |
|     visits     |
| water_quality  |
|  water_source  |
| well_pollution |
|    location    |
+----------------+

We can see that there are only **6** tables related to each other as per the `data_dictionary` table. Great, with the `data_dictionary` table as our map and the `md_water_services` database as our landscape, we now know how to navigate our data landscape. We just go ahead and view the first fiew rows for every table save for the `data_dictionary` table as we already know that it is more of a reference point for our real data in the database. You can run the query below multiple times while changing the table name after the `FROM` clause and it should display the first 10 records and each of their attributes per table/entity

In [7]:
%sql SELECT * FROM employee;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

+----------------------+----------------+---------------+-------+---------------------------+---------------+-----------+----------------+
| assigned_employee_id | employee_name  |  phone_number | email |          address          | province_name | town_name |    position    |
+----------------------+----------------+---------------+-------+---------------------------+---------------+-----------+----------------+
|          0           |  Amara Jengo   | +99637993287  |  None |  36 Pwani Mchangani Road  |     Sokoto    |   Ilanga  | Field Surveyor |
|          1           |  Bello Azibo   | +99643864786  |  None |   129 Ziwa La Kioo Road   |    Kilimani   |   Rural   | Field Surveyor |
|          2           |  Bakari Iniko  | +99222599041  |  None |   18 Mlima Tazama Avenue  |    Hawassa    |   Rural   | Field Surveyor |
|          3           | Malachi Mavuso | +99945849900  |  None |     100 Mogadishu Road    |     Akatsi    |   Lusaka  | Field Surveyor |
|          4           |  Cheche Buhle  | +99381679640  |  None |      1 Savanna Street     |     Akatsi    |   Rural   | Field Surveyor |
|          5           | Zuriel Matembo | +99034075111  |  None |  26 Bahari Ya Faraja Road |    Kilimani   |   Rural   | Field Surveyor |
|          6           |  Deka Osumare  | +99379364631  |  None |    104 Kenyatta Street    |     Akatsi    |   Rural   | Field Surveyor |
|          7           | Lalitha Kaburi | +99681623240  |  None | 145 Sungura Amanpour Road |    Kilimani   |   Rural   | Field Surveyor |
|          8           |  Enitan Zuri   | +99248509202  |  None |      117 Kampala Road     |    Hawassa    |  Zanzibar | Field Surveyor |
|          10          |   Farai Nia    | +99570082739  |  None | 33 Angélique Kidjo Avenue |     Amanzi    |   Dahabu  | Field Surveyor |
+----------------------+----------------+---------------+-------+---------------------------+---------------+-----------+----------------+
Truncated to displaylimit of 10.

## Diving Into the Water Sources

Now that we are familiar with what each entity in our database entails, we can dive deeper into specific aspects of our database. A good starting point is understanding the types of water sources recorded in the database. To get that information, we can inspect the `water_source` table

In [8]:
%sql SELECT DISTINCT type_of_water_source FROM md_water_services.water_source;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+----------------------+
| type_of_water_source |
+----------------------+
|     tap_in_home      |
|  tap_in_home_broken  |
|         well         |
|      shared_tap      |
|        river         |
+----------------------+

We can see that we have **5** unique types of water sources recorded in our database. Understanding what each of these types mean is paramount to deciphering proper data-driven decision making reports.

## Unpacking the Visits to Water Sources

The `visits` entity in the database logs information on each water source each and every time the water source is visited. From our data exploration above, we also noticed that there is a `time_in_queue` attribute in this entity. Let's experiment and retreive records from this entity `WHERE time_in_queue > 500`.

In [9]:
%%sql 
SELECT * 
FROM md_water_services.visits 
WHERE time_in_queue > 500 
ORDER BY time_in_queue DESC;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

105 rows affected.

+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
| record_id | location_id |  source_id   |    time_of_record   | visit_count | time_in_queue | assigned_employee_id |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
|   30007   |  AmRu14612  | AmRu14612224 | 2022-04-02 08:55:00 |      2      |      539      |          8           |
|   51858   |  HaRu19538  | HaRu19538224 | 2023-03-04 18:04:00 |      3      |      539      |          4           |
|   53278   |  AkRu05704  | AkRu05704224 | 2023-03-25 13:48:00 |      2      |      539      |          36          |
|   45317   |  HaRu20126  | HaRu20126224 | 2022-11-19 14:22:00 |      6      |      538      |          16          |
|   57408   |  SoRu35388  | SoRu35388224 | 2023-05-27 08:52:00 |      5      |      538      |          1           |
|   20372   |  KiZu31117  | KiZu31117224 | 2021-11-06 09:37:00 |      3      |      537      |          10          |
|   33650   |  KiRu29348  | KiRu29348224 | 2022-05-28 12:58:00 |      2      |      537      |          10          |
|   31310   |  SoRu37865  | SoRu37865224 | 2022-04-23 06:01:00 |      2      |      535      |          40          |
|   38947   |  SoRu38095  | SoRu38095224 | 2022-08-13 13:48:00 |      6      |      535      |          30          |
|   52264   |  HaRu17383  | HaRu17383224 | 2023-03-11 07:10:00 |      5      |      535      |          30          |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
Truncated to displaylimit of 10.

We can further investigate the `type_of_water_source` with such long `time_in_queue`. to do this, let's select the first three `source_id`s from the `visits` entity and search for them in the `water_source` entity

In [10]:
%%sql
SELECT 
    source_id,
    type_of_water_source,
    number_of_people_served
FROM water_source
WHERE source_id IN ("AmRu14612224", "HaRu19538224", "AkRu05704224");

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

3 rows affected.

+--------------+----------------------+-------------------------+
|  source_id   | type_of_water_source | number_of_people_served |
+--------------+----------------------+-------------------------+
| AkRu05704224 |      shared_tap      |           3398          |
| AmRu14612224 |      shared_tap      |           3118          |
| HaRu19538224 |      shared_tap      |           3142          |
+--------------+----------------------+-------------------------+

We can see that these are water sources of the type `shared_tap` serving more than **3000** people. Keep in mind that from the information in the project description that there were other sources that were visited more than once by the surveyors to see if there was a change in `time_in_queue`.

## Assessing the Quality of Water Sources

Now that we've explored the visits made to various water sources, we can narrow down our analysis on the quality of water sources. The quality of water sources is the whole point of the survey. Recall that we have an entity in our database that has records pertaining to `water_quality`. This entity has recorded water quality scores assigned by surveyors.

In [11]:
%%sql
# Retrieve all records in with good water sources visited twice
SELECT *
FROM water_quality
WHERE subjective_quality_score = 10
AND visit_count = 2;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

218 rows affected.

+-----------+--------------------------+-------------+
| record_id | subjective_quality_score | visit_count |
+-----------+--------------------------+-------------+
|     59    |            10            |      2      |
|    137    |            10            |      2      |
|    269    |            10            |      2      |
|    363    |            10            |      2      |
|    378    |            10            |      2      |
|    618    |            10            |      2      |
|    752    |            10            |      2      |
|    801    |            10            |      2      |
|    819    |            10            |      2      |
|    850    |            10            |      2      |
+-----------+--------------------------+-------------+
Truncated to displaylimit of 10.

According to the project's narrative, the surveyors only revisited shared taps and not other types of water sources. So technically speaking, there shouldn't records with a visit count of more than or equal to **2** where there are good water sources indicated by a `subjective_quality_score` of **10**. So why are there **218** records with a `visit_count` of **2** with a `subjective_quality_score` of **10**?

## Investigating any Pollution Issues

Finally, let's investigate the pollution issues as per Maji Ndogo's database. There's a table with recorded pollution/contamination data of wells in Maji Ndogo. Let's have a quick look at that table.

In [12]:
%sql SELECT * FROM well_pollution;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

17383 rows affected.

+--------------+---------------------+----------------------------------------------------------+---------------+-------------+--------------------------+
|  source_id   |         date        |                       description                        | pollutant_ppm |  biological |         results          |
+--------------+---------------------+----------------------------------------------------------+---------------+-------------+--------------------------+
| KiRu28935224 | 2021-01-04 09:17:00 |                Bacteria: Giardia Lamblia                 |      0.0      |   495.898   | Contaminated: Biological |
| AkLu01628224 | 2021-01-04 09:53:00 |                    Bacteria: E. coli                     |      0.0      |   6.09608   | Contaminated: Biological |
| HaZa21742224 | 2021-01-04 10:37:00 |    Inorganic contaminants: Zinc, Zinc, Lead, Cadmium     |     2.715     |     0.0     |  Contaminated: Chemical  |
| HaRu19725224 | 2021-01-04 11:04:00 |                          Clean                           |   0.0288593   | 9.56996e-05 |          Clean           |
| SoRu35703224 | 2021-01-04 11:29:00 |                    Bacteria: E. coli                     |      0.0      |   22.5009   | Contaminated: Biological |
| AkHa00070224 | 2021-01-04 11:42:00 |             Inorganic contaminants: Cadmium              |    5.46739    |     0.0     |  Contaminated: Chemical  |
| HaSe21346224 | 2021-01-04 11:52:00 |                          Clean                           |   0.0140376   | 8.98989e-05 |          Clean           |
| HaYa21468224 | 2021-01-04 12:03:00 | Inorganic contaminants: Chromium, Barium, Chromium, Lead |    6.05137    |     0.0     |  Contaminated: Chemical  |
| SoRu36278224 | 2021-01-04 12:24:00 |                Parasite: Cryptosporidium                 |      0.0      |   485.162   | Contaminated: Biological |
| AkLu02155224 | 2021-01-04 12:29:00 |        Inorganic contaminants: Selenium, Arsenic         |    7.64106    |     0.0     |  Contaminated: Chemical  |
+--------------+---------------------+----------------------------------------------------------+---------------+-------------+--------------------------+
Truncated to displaylimit of 10.

We can see that by viewing the `results` column, some wells are `Clean` while some are `Contaminated: Biological`ly while others are `Contaminated: Chemical`ly. Each of the records has a `source_id` included so this can help us link the `results` to the sources in other tables in the database.

As per the project description, the `description` column is text recorded down by scientists which will be challenging to process. The `biological` column is numeric and measured in the units `CFU/mL` which is the measure of how much contamination is in the water. **0** is clean but anything more than **0.01** is contaminated.

We can check the integrity of the data/records just to make sure we don't have any **false positives** which might mislead people to consume contaminated, unsafe water.

In [13]:
%%sql
SELECT * 
FROM well_pollution
WHERE results = "Clean"
AND biological > 0.01;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

64 rows affected.

+--------------+---------------------+---------------------------------+---------------+------------+---------+
|  source_id   |         date        |           description           | pollutant_ppm | biological | results |
+--------------+---------------------+---------------------------------+---------------+------------+---------+
| AkRu08936224 | 2021-01-08 09:22:00 |        Bacteria: E. coli        |   0.0406458   |  35.0068   |  Clean  |
| AkRu06489224 | 2021-01-10 09:44:00 | Clean Bacteria: Giardia Lamblia |   0.0897904   |   38.467   |  Clean  |
| SoRu38011224 | 2021-01-14 15:35:00 |        Bacteria: E. coli        |   0.0425095   |  19.2897   |  Clean  |
| AkKi00955224 | 2021-01-22 12:47:00 |        Bacteria: E. coli        |   0.0812092   |  40.2273   |  Clean  |
| KiHa22929224 | 2021-02-06 13:54:00 |        Bacteria: E. coli        |   0.0722537   |  18.4482   |  Clean  |
| KiRu25473224 | 2021-02-07 15:51:00 | Clean Bacteria: Giardia Lamblia |   0.0630094   |  24.4536   |  Clean  |
| HaRu17401224 | 2021-03-01 13:44:00 | Clean Bacteria: Giardia Lamblia |   0.0649209   |  25.8129   |  Clean  |
| AkRu07137224 | 2021-03-04 13:41:00 | Clean Bacteria: Giardia Lamblia |   0.0656843   |  18.2978   |  Clean  |
| KiRu27205224 | 2021-03-13 14:17:00 | Clean Bacteria: Giardia Lamblia |   0.0418018   |  49.4281   |  Clean  |
| AkLu02307224 | 2021-03-13 15:41:00 |        Bacteria: E. coli        |   0.0709682   |   35.203   |  Clean  |
+--------------+---------------------+---------------------------------+---------------+------------+---------+
Truncated to displaylimit of 10.

It seems like, in some cases, if the `description` field begins with the word "Clean", the results have been classified as "Clean" in the `results` column, even though the `biological` column is **> 0.01**. Let's dive deeper into the cause of the issue with the biological contamination data.

As per the project specifications, `description`s should only have the word "Clean" if there is **no biological contamination (and no chemical pol-lutants)**. This means that we need to find and remove the "Clean" part from all the `description`s that do have a biological contamination so this mistake is not made again.

The second issue has arisen from this error, but it is much more problematic. Some wells have been marked as "Clean" in the `results` column because the `description` had the word "Clean" in it, even though they have a biological contamination. So we need to find all the results that have a value **greater than 0.01** in the `biological` column and have been set to "Clean" in the `results` column.

First, let's look at the `description`s. We need to identify the records that mistakenly have the word "Clean" in the `description`. However, it is important to remember that not all of the field surveyors used the description to set the results - some checked the actual data.

In [14]:
%%sql
# Retrieve all records with errornous descriptions
SELECT *
FROM well_pollution
WHERE description LIKE "Clean_%";

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

38 rows affected.

+--------------+---------------------+---------------------------------+---------------+------------+---------+
|  source_id   |         date        |           description           | pollutant_ppm | biological | results |
+--------------+---------------------+---------------------------------+---------------+------------+---------+
| AkRu06489224 | 2021-01-10 09:44:00 | Clean Bacteria: Giardia Lamblia |   0.0897904   |   38.467   |  Clean  |
| KiRu25473224 | 2021-02-07 15:51:00 | Clean Bacteria: Giardia Lamblia |   0.0630094   |  24.4536   |  Clean  |
| HaRu17401224 | 2021-03-01 13:44:00 | Clean Bacteria: Giardia Lamblia |   0.0649209   |  25.8129   |  Clean  |
| AkRu07137224 | 2021-03-04 13:41:00 | Clean Bacteria: Giardia Lamblia |   0.0656843   |  18.2978   |  Clean  |
| KiRu27205224 | 2021-03-13 14:17:00 | Clean Bacteria: Giardia Lamblia |   0.0418018   |  49.4281   |  Clean  |
| AkHa00514224 | 2021-04-11 12:11:00 | Clean Bacteria: Giardia Lamblia |   0.0305404   |  22.0255   |  Clean  |
| AmAm09776224 | 2021-05-23 11:28:00 | Clean Bacteria: Giardia Lamblia |   0.0963821   |  13.6574   |  Clean  |
| SoIl32894224 | 2021-07-11 11:37:00 | Clean Bacteria: Giardia Lamblia |   0.0712408   |  5.44957   |  Clean  |
| AkRu07366224 | 2021-07-23 11:19:00 | Clean Bacteria: Giardia Lamblia |   0.0969458   |  26.0308   |  Clean  |
| KiHa23443224 | 2021-09-05 12:34:00 | Clean Bacteria: Giardia Lamblia |     0.0828    |  13.7162   |  Clean  |
+--------------+---------------------+---------------------------------+---------------+------------+---------+
Truncated to displaylimit of 10.

The query returned **38** wrong descriptions. Now we need to fix these descriptions so that we don't encounter this issue again in the future. Looking at the results we can see two different descriptions that we need to fix:
1. All records that mistakenly have Clean Bacteria: E. coli should updated to Bacteria: E. coli
2. All records that mistakenly have Clean Bacteria: Giardia Lamblia should updated to Bacteria: Giardia Lamblia

The second issue we need to fix is in our `results` column. We need to update the `results` column from "Clean" to "Contaminated: Biological" where the `biological` column has a value **greater than 0.01**.

> **NOTE**: The query below 👇🏾 should only be run once as the changes will persist through the database, so keep that in mind when running queries in the notebook environment incase you restart the kernel and run all cells.

In [15]:
%%sql
# Update all the erronous values in description and results attributes
UPDATE
    well_pollution
SET
    description = "Bacteria: E. coli"
WHERE
    description = "Clean Bacteria: E. coli";

UPDATE
    well_pollution
SET
    description = "Bacteria: Giardia Lamblia"
WHERE
    description = "Clean Bacteria: Giardia Lamblia";

UPDATE
    well_pollution
SET
    results = "Contaminated: Biological"
WHERE
    biological > 0.01 AND results = "Clean";

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

26 rows affected.

12 rows affected.

64 rows affected.

++
||
++
++